### ISSUE TITLE
Breaking equivariance

### ISSUE THREAD

**Issue by brennanaba** [lucidrains/se3-transformer-pytorch#3]
Hi, Thanks a lot for your work!!

I was running some equivariance tests on your implementation of the SE3-Transformer and found that it is not always conserved. It does not break every time and unfortunately I do not know where the bug is. 

I have appended an image with an example of equivariance not being conserved. 

![image](https://user-images.githubusercontent.com/56822523/107859610-78db7300-6e32-11eb-8cd2-268c0e8f7955.png)

#### To generate this image I used the following code:

        import numpy as np
        import matplotlib.pyplot as plt
        import torch
        
        from se3_transformer_pytorch import SE3Transformer

Make some data

        zline = np.arange(0, 2, 0.05)
        xline = np.sin(zline * 2 * np.pi) 
        yline = np.cos(zline * 2 * np.pi)
        points = np.array([xline, yline, zline])
        geom = torch.tensor(points.transpose())[None,:].float()
        feat = torch.randint(0, 20, (1, geom.shape[1],1)).float()
        
        def rot_matrix(x):
            # Rotation matrix
            a ,b ,c = 2*np.pi*x
            return np.array([
                [np.cos(a)*np.cos(b), np.cos(a)*np.sin(b)*np.sin(c)- np.sin(a)*np.cos(c), np.cos(a)*np.sin(b)*np.cos(c)+ np.sin(a)*np.sin(c)],
                [np.sin(a)*np.cos(b), np.sin(a)*np.sin(b)*np.sin(c)+ np.cos(a)*np.cos(c), np.sin(a)*np.sin(b)*np.cos(c)- np.cos(a)*np.sin(c)],
                [-np.sin(b)         , np.cos(b)*np.sin(c)                               , np.cos(b)*np.cos(c)                               ]
            ])

Initialize model

        mdl = SE3Transformer(
            dim = 1,
            depth = 3,
            input_degrees = 1,
            num_degrees = 2,
            output_degrees = 2,
            reduce_dim_out = True,
        )
        
        def model(geom,feat):
            return geom + mdl(feat,geom, return_type = 1)

Check Rotation Invariance:

        with torch.no_grad():
            
            Q = torch.tensor(rot_matrix(np.random.random(3))).float()
            prerotated = model(geom @ Q, feat).squeeze().detach().numpy().transpose()
            posrotated = (model(geom, feat) @ Q).squeeze().detach().numpy().transpose()
        
            fig = plt.figure(dpi = 200)
            ax = plt.axes(projection="3d")
        
            ax.plot3D(prerotated[0], prerotated[1], prerotated[2], "r", linewidth=1.1)
            ax.plot3D(posrotated[0], posrotated[1], posrotated[2], "b", linewidth=0.5)
        
            plt.legend(["Pre-Rotated", "Post-Rotated"])
            plt.show()

Check Translation Invariance:

        with torch.no_grad():
            
            x0 = 1*torch.rand(3)
            prerotated = model(geom + x0, feat).squeeze().detach().numpy().transpose()
            posrotated = (model(geom, feat) + x0).squeeze().detach().numpy().transpose()
        
            fig = plt.figure(dpi = 200)
            ax = plt.axes(projection="3d")
        
            ax.plot3D(prerotated[0], prerotated[1], prerotated[2], "r", linewidth=1.1)
            ax.plot3D(posrotated[0], posrotated[1], posrotated[2], "b", linewidth=0.5)
        
            plt.legend(["Pre-Translated", "Post-Translated"])
            plt.show()

Hope this helps!!

**Comment by lucidrains** [comment]
@brennanaba Thanks Brennan for reporting this! Do you happen to know if the same issue does or does not exist for the official repository that Fabian put out?

**Comment by lucidrains** [comment]
@brennanaba Could you try the latest version? (0.0.14) I turned off an experimental feature I had on by accident (but it may or may not be the culprit)

**Comment by brennanaba** [comment]
Hi @lucidrains, thanks for answering so quickly. This does not happen on the version that Fabian put out. I will give the new version a try later today and let you know.

**Comment by brennanaba** [comment]
I am still getting the same error on version 0.0.14. 

![image](https://user-images.githubusercontent.com/56822523/107872714-0acf9400-6ea4-11eb-9886-ef06f066fd56.png)

Just to clarify, this only happens around once for every ten runs of the code. So for some reason it seem to depends on how the model is initialised. Sorry I can't be of more help!!

**Comment by lucidrains** [comment]
@brennanaba No, thank you! :) As long as you can confirm Fabian's code is alright, then I can hunt down the bug. I'll put both into deterministic mode, feed both the same input until the output differs and find out where the offending code is

**Comment by lucidrains** [comment]
@brennanaba I did some more tinkering, and it seems like it is a precision problem

```python
import torch
from se3_transformer_pytorch.se3_transformer_pytorch import SE3Transformer
from se3_transformer_pytorch.irr_repr import rot
from se3_transformer_pytorch.utils import torch_default_dtype

iters = 100

dtype = torch.float64
# dtype = torch.float32

with torch_default_dtype(dtype):
    dim = 4
    max_error = 0

    model = SE3Transformer(
        dim = dim,
        depth = 3,
        input_degrees = 1,
        num_degrees = 2,
        output_degrees = 2,
        reduce_dim_out = True,
    ).cuda()

    for i in range(iters):
        feats = torch.randn(1, 16, dim, dtype = dtype).cuda()
        coors = torch.randn(1, 16, 3, dtype = dtype).cuda()
        mask  = torch.ones(1, 16).bool().cuda()

        R   = rot(15, 0, 45).type(dtype).cuda()
        out1 = model(feats, coors @ R, mask, return_type = 1)
        out2 = model(feats, coors, mask, return_type = 1) @ R

        diff = (out1 - out2).max()
        print(f'{i} out of {iters} - {diff:.8f}')
        if diff > max_error:
            max_error = diff

    print('max error is', max_error)
```

In 100 tries, the maximum error I get with float64 is `1.5e-7`

For float32, I hit `1e-4` :(

You can try this by running the script above yourself

**Comment by lucidrains** [comment]
oh also, upgrade to `0.0.15` in order for the model to work with float64 inputs, sorry, work in progress :pray:

**Comment by lucidrains** [comment]
my guess is that there is some approximation that leads to unstable results in the context of 'low' precision

**Comment by lucidrains** [comment]
in either case, it seems like doing everything in float64 'solves' the issue

**Comment by lucidrains** [comment]
ran float64 for 5000 times (each time calculates new basis vectors, unlike DGL graph batching), and the maximum error is `3.2820e-06`

**Comment by lucidrains** [comment]
@brennanaba I'll see if I can bring in reversible networks https://arxiv.org/abs/1707.04585 and perhaps we can just use float64 throughout the network without having to incur a great cost

**Comment by lucidrains** [comment]
@brennanaba do you mind if I include your code as a colab notebook for others to try on the Readme?

**Comment by brennanaba** [comment]
@lucidrains Sorry for the late reply!! Feel free to share my code, I will check the new version tomorrow and let you know.

**Comment by brennanaba** [comment]
Hi Lucid, works like a charm with float64. I wonder at which step the network requires such high precision. Thanks a lot for fixing this!!

**Comment by lucidrains** [comment]
@brennanaba No problem! I'll step through the network this week to figure out just where in the network it needs to be high precision. Would be disappointing if it were to be all of it, but I can always pull the reversible net card.

**Comment by lucidrains** [comment]
@brennanaba No problem! I'll step through the network this week to figure out just where in the network it needs to be high precision. Would be disappointing if it were to be all of it, but I can always pull the reversibiliy card.

**Comment by lucidrains** [comment]
There is also the possibility I introduced a bug that makes the network more prone to precision errors, but I cannot imagine how that could be the case

**Comment by brennanaba** [comment]
Hi Lucid!! Just as an fyi and because you have it on your Readme. 
If you try and predict coordinates directly by using the SE3-Transformer (like for you Alphafold2 example), you end up breaking translational equivariance. This is because the coordinates predicted will have to depend on your origin being fixed in space. This is the reason I force the model to predict  "correction" vectors for each input coordinate in the example I sent you. 

To put it in coding terms, this breaks equivariance:

    import torch
    from se3_transformer_pytorch.se3_transformer_pytorch import SE3Transformer
    from se3_transformer_pytorch.irr_repr import rot
    from se3_transformer_pytorch.utils import torch_default_dtype
    
    iters = 100
    
    dtype = torch.float64
    #dtype = torch.float32
    
    with torch_default_dtype(dtype):
        dim = 4
        max_error = 0
    
        model = SE3Transformer(
            dim = dim,
            depth = 3,
            input_degrees = 1,
            num_degrees = 2,
            output_degrees = 2,
            reduce_dim_out = True,
        )
    
        for i in range(iters):
            feats = torch.randn(1, 16, dim, dtype = dtype)
            coors = torch.randn(1, 16, 3, dtype = dtype)
            mask  = torch.ones(1, 16).bool()
    
            x0 = torch.rand(3)
            out1 = model(feats, coors + x0, mask, return_type = 1)
            out2 = model(feats, coors, mask, return_type = 1) + x0
    
            diff = (out1 - out2).max()
            print(f'{i} out of {iters} - {diff:.8f}')
            if diff > max_error:
                max_error = diff
    
        print('max error is', max_error)

Hope that helps!!

**Comment by lucidrains** [comment]
@brennanaba duly noted! i've updated the readme!
